### PR TITLE
pkg/bpftrace: Add bpftrace as a core pkg

### PR DESCRIPTION
**Pull request by simar7**
**- What I did**
This PR adds [bpftrace](https://github.com/iovisor/bpftrace) as a package to linuxkit. It includes all the necessary components and directory structures in order to build a bpftrace inside of a docker container.

Addresses: https://github.com/linuxkit/linuxkit/issues/3201

**- How I did it**
I started off by looking at existing packages for a reference model of how to structure a package in linuxkit. I also read the https://github.com/linuxkit/linuxkit/blob/master/docs/packages.md doc.

**- How to verify it**
You can build it locally to test it out using the following command:
```linuxkit pkg build -org=<orgname> -disable-content-trust <path/to/bpftrace>```

**- Description for the changelog**

This commit adds bpftrace to linuxkit packages and addresses the issue: https://github.com/linuxkit/linuxkit/issues/3201

**- A picture of a cute animal (not mandatory but encouraged)**
image

Signed-off-by: Simarpreet Singh <simar@linux.com>

### CODE REVIEW

File: pkg/bpftrace/build-release.sh [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
```
Comment (justincormack) [inline pkg/bpftrace/build-release.sh]: can you use /bin/sh - this does not use any bash features...

File: pkg/bpftrace/build.sh [hunk 2]
```diff
@@ -0,0 +1,14 @@
+set -e
```
Comment (justincormack) [inline pkg/bpftrace/build.sh]: this is missing the hashbang...

**Comment by justincormack** [comment]
I don't think we should vendor the bpftrace source here, can we download from a specific git commit instead?

**Comment by rn** [comment]
Thanks for importing, this. But as @justincormack said, could you please follow a pattern similar to the other packages which have a `Dockerfile` which check out a specific revision of the source code and then compile it. If you look at the `bpftrace` `Dockerfile` is basically just installs the too chain, which should already be present in the `linuxkit/alpine` package, and then calls `build.sh`

File: pkg/bpftrace/build.sh (hunk 2 above)
Comment (simar7) [reply pkg/bpftrace/build.sh]: Thanks. Fixed in the next commit.

File: pkg/bpftrace/build-release.sh (hunk 1 above)
Comment (simar7) [reply pkg/bpftrace/build-release.sh]: Thanks. Fixed in the next commit.

**Comment by simar7** [comment]
@rn and @justincormack : Good point on the vendoring out the bpftrace files. I've modified the Dockerfile to checkout from source.

File: pkg/bpftrace/build-release.sh [hunk 3]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
```
Comment (rn) [inline pkg/bpftrace/build-release.sh]: this should no longer be needed

File: pkg/bpftrace/build.sh [hunk 4]
```diff
@@ -0,0 +1,15 @@
+#!/bin/sh
```
Comment (rn) [inline pkg/bpftrace/build.sh]: this should no longer be needed

File: pkg/bpftrace/build.yml [hunk 5]
```diff
@@ -0,0 +1,13 @@
+image: bpftrace
+network: true
+arches: 
+  - amd64
+gitrepo: github.com/iovisor/bpftrace
```
Comment (rn) [inline pkg/bpftrace/build.yml]: this is not needed.

File: pkg/bpftrace/build.yml [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+config:
+  pid: host
+  binds:
+    - /proc:/host/proc
+    - /sys:/host/sys
+    - /:/rootfs
```
Comment (rn) [inline pkg/bpftrace/build.yml]: why is this here?

File: pkg/bpftrace/build.yml (hunk 5 above)
Comment (simar7) [reply pkg/bpftrace/build.yml]: I can remove this. It was part of the packaging readme https://github.com/linuxkit/linuxkit/blob/master/docs/packages.md#package-source so I added.

Comment (rn) [reply pkg/bpftrace/build.yml]: I think this refers to the location of the `build.yml` and defaults to the linuxkit repository so can be omited

File: pkg/bpftrace/build-release.sh (hunk 3 above)
Comment (simar7) [reply pkg/bpftrace/build-release.sh]: Removed.

File: pkg/bpftrace/build.yml (hunk 5 above)
Comment (simar7) [reply pkg/bpftrace/build.yml]: Gotcha. Fixed in the next commit.

File: pkg/bpftrace/build.yml (hunk 6 above)
Comment (simar7) [reply pkg/bpftrace/build.yml]: copy/pasta. Removed.

File: pkg/bpftrace/build.sh (hunk 4 above)
Comment (simar7) [reply pkg/bpftrace/build.sh]: Removed.

**Comment by simar7** [comment]
Squashed into a single commit and force pushed with all the recommendations.

File: pkg/bpftrace/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+ENV COMMIT=8f7f8214d7dd7bc25b7740a3c0e9a580a89e0244
+ENV STATIC_LINKING=on
+ENV RUN_TESTS=0
+
+RUN git clone ${GITREPO} /bpftrace \
+    && cd /bpftrace/ \ 
+    && mkdir -p /bpftrace/build \
+    && git checkout ${COMMIT}
+
+ENTRYPOINT ["/bin/sh", "/bpftrace/docker/build.sh", "/bpftrace/build", "Release" ]
```
Comment (tjfontaine) [inline pkg/bpftrace/Dockerfile]: Should this really be an `ENTRYPOINT` or another command in the previous `RUN`? I think it's the latter.

And then if it is in fact a fully static binary, then the final stanzas can be

```
FROM scratch
WORKDIR
ENTRYPOINT[]
## Assuming the real value of the path to bpftrace
COPY --from=build /bpftrace/build/bpftrace /bpftrace
CMD ["/bpftrace"]
```

and the first can be `FROM linuxkit/alpine:ace75d0ec6978762d445083d6c8c6336bdb658ed AS build`

Comment (simar7) [reply pkg/bpftrace/Dockerfile]: Great suggestion. I'll try playing around with this.

Comment (simar7) [reply pkg/bpftrace/Dockerfile]: I've implemented this in the next commit.

**Comment by simar7** [comment]
Although this now runs as expected, I'm curious if this means we'd need a custom kernel with appropriate debug capabilities. For instance this is an example run that should've listed some probes.

```docker run -it bpftrace -l

No such file or directory: /sys/kernel/debug/tracing/available_filter_functions
```

File: test/cases/040_packages/002_bpftrace/test.yml [hunk 8]
```diff
@@ -0,0 +1,29 @@
+kernel:
+  image: linuxkit/kernel:4.14.81
+  cmdline: "console=ttyS0 console=ttyAMA0"
+init:
+  - linuxkit/init:v0.5
+  - linuxkit/runc:v0.5
+  - wombat/bpftrace:1a2cc3766bf1f12090a3f8ea584d141f0ec860c0
```
[the remark this answers is not in the document]
Comment (simar7) [reply test/cases/040_packages/002_bpftrace/test.yml]: Any thoughts on this cc @justincormack @rn @tjfontaine

**Comment by rn** [comment]
@simar7 this looks good to me. Thanks for setting this up.

To add a new package, one of the maintainers has to add this on hub and set up signing on the notary server. A maintainer also has to build and push the new package for the CI to pass.

Unfortunately, I hadn't had time earlier this week and am travelling next week. I see if I can do the first bit tomorrow morning

File: pkg/bpftrace/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+ENV COMMIT=8f7f8214d7dd7bc25b7740a3c0e9a580a89e0244
+ENV STATIC_LINKING=on
+ENV RUN_TESTS=0
+
+RUN git clone ${GITREPO} /bpftrace \
+    && cd /bpftrace/ \ 
+    && mkdir -p /bpftrace/build \
+    && git checkout ${COMMIT} \
+    && chmod +x /bpftrace/docker/build.sh \
+    && /bpftrace/docker/build.sh /bpftrace/build Release
```
Comment (tjfontaine) [inline pkg/bpftrace/Dockerfile]: While testing this, I noticed that there's no parallelism done automatically, maybe add a `-j$(nproc)` or similar here? The `build.sh` seems to support arbitrary `make` arguments subsequent to `Release`

**Comment by tjfontaine** [comment]
This pkg seems to work, I haven't tried running the linuxkit test though

```
(ns: getty) linuxkit-3efcfc3734a0:~# /proc/1/root/bpftrace -e 'tracepoint:raw_syscalls:sys_enter { @[comm] = count(); }'
chdir(/lib/modules/4.14.81-linuxkit/build): No such file or directory
definitions.h:1:10: fatal error: 'linux/types.h' file not found
Attaching 1 probe...


@[bpftrace]: 6
@[tini]: 30
@[containerd-shim]: 148
@[containerd]: 799


(ns: getty) linuxkit-3efcfc3734a0:~# /proc/1/root/bpftrace -l | wc -l
38542
```

**Comment by tjfontaine** [comment]
Ok, I took a few laps testing this today, and have some suggestions --

```diff
diff --git a/pkg/bpftrace/Dockerfile b/pkg/bpftrace/Dockerfile
index 207345449..88c409cd1 100644
--- a/pkg/bpftrace/Dockerfile
+++ b/pkg/bpftrace/Dockerfile
@@ -27,11 +27,15 @@ RUN git clone ${GITREPO} /bpftrace \
     && cd /bpftrace/ \ 
     && mkdir -p /bpftrace/build \
     && git checkout ${COMMIT} \
-    && chmod +x /bpftrace/docker/build.sh \
-    && /bpftrace/docker/build.sh /bpftrace/build Release
+    && cd /bpftrace/build \
+    && export MAKEFLAGS=-j$(nproc) \
+    && cmake -DCMAKE_BUILD_TYPE="Release" -DSTATIC_LINKING:BOOL=ON -DRUN_TESTS=0 -DCMAKE_INSTALL_PREFIX=/usr ../ \
+    && make -j$(nproc) \
+    && DESTDIR=/out make install \
+    && chmod +x /out/usr/share/bpftrace/tools/*.bt
 
 FROM scratch
 WORKDIR /
-COPY --from=build /bpftrace/build/src/bpftrace /bpftrace
-ENTRYPOINT [ "/bpftrace" ]
-CMD ["/bpftrace", "-h"]
\ No newline at end of file
+COPY --from=build /out/ /
+ENTRYPOINT [ "/usr/bin/bpftrace" ]
+CMD ["/usr/bin/bpftrace", "-h"]
```

Essentially this just implements the `build.sh` inline to be able to pass more flags to `cmake` than what is currently available [there may be a more elegant `cmake` approach for this]. The key here, is we're setting the prefix to `/usr` so the tools look in more accurate platform locations for things, like `/usr/include` etc etc, also it uses `make install DESTDIR=/out` which allows us to ship the tools that are shipped with `bpftrace` as well.

As a result, you can either `nsenter -t1 -a -- bpftrace` or `export PATH=$PATH:/proc/1/root/usr/bin` though the latter is slightly "dangerous", assuming it is on your `$PATH` though, you can use the tools by either copying them into `getty`/`ssh` from the root namespace, or using them in place `/proc/1/root/usr/share/bpftrace/tools/pidpersec.bt`

also, using an image like

```yml
kernel:
  image: linuxkit/kernel:4.14.81
  cmdline: "console=tty0 console=ttyS0 console=ttyAMA0 console=ttysclp0"
init:
  - linuxkit/init:c563953a2277eb73a89d89f70e4b6dcdcfebc2d1
  - linuxkit/runc:83d0edb4552b1a5df1f0976f05f442829eac38fe
  - linuxkit/containerd:326b096cd5fbab0f864e52721d036cade67599d6
  - linuxkit/ca-certificates:v0.6
  - linuxkit/bpftrace:1a2cc3766bf1f12090a3f8ea584d141f0ec860c0-dirty
  - linuxkit/kernel:4.14.81
onboot:
  - name: sysctl
    image: linuxkit/sysctl:v0.6
  - name: dhcpcd
    image: linuxkit/dhcpcd:v0.6
    command: ["/sbin/dhcpcd", "--nobackground", "-f", "/dhcpcd.conf", "-1"]
services:
  - name: getty
    image: linuxkit/getty:2eb742cd7a68e14cf50577c02f30147bc406e478
    env:
     - INSECURE=true
  - name: rngd
    image: linuxkit/rngd:v0.6
trust:
  org:
    - linuxkit
```

You're able to `tar xf /proc/1/root/kernel-headers.tar` such that you can avoid `bpftrace`'s invocation of `clang` from failing, and it can be in sync with the kernel version you're running.

These are just some suggestions, I'll file some tickets with `bpftrace` directly regarding some of the pitfalls of their implementation, but that's outside the scope of this PR.

**Comment by simar7** [comment]
@tjfontaine those are some good suggestions but I feel as an initial implementation of this PR it's best to keep it small to start and later on we can improve upon this. Thoughts?

**Comment by tjfontaine** [comment]
I'm fine with whatever the maintainers decide, I was just testing it out :)

**Comment by simar7** [comment]
hey @justinbatcf and @rn sorry to nudge but if you folks could see if this PR makes the cut for a first thin slice implementation, then we could merge it in.

File: pkg/bpftrace/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+RUN apk add --update \
+    bison \
+    build-base \
+    clang-dev \
+    clang-static \
+    cmake \
+    elfutils-dev \
+    flex-dev \
+    git \
+    linux-headers \
```
Comment (rn) [inline pkg/bpftrace/Dockerfile]: This compiles it against the headers from Alpine Linux but not the LinuxKit kernels. Is that a problem?

File: pkg/bpftrace/Dockerfile (hunk 9 above)
Comment (rn) [reply pkg/bpftrace/Dockerfile]: Adding some parallelism here would be good. It seems to take some time to compile...

File: pkg/bpftrace/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    && cd /bpftrace/ \ 
+    && mkdir -p /bpftrace/build \
+    && git checkout ${COMMIT} \
+    && chmod +x /bpftrace/docker/build.sh \
+    && /bpftrace/docker/build.sh /bpftrace/build Release
+
+FROM scratch
+WORKDIR /
+COPY --from=build /bpftrace/build/src/bpftrace /bpftrace
+ENTRYPOINT [ "/bpftrace" ]
```
Comment (rn) [inline pkg/bpftrace/Dockerfile]: I would remove this and the next line. The `bpftrace` is unlikely to be used as a stand alone package running in a container. to be consistent with other package file make this:
```
FROM scratch
ENTRYPOINT []
CMD []
WORKDIR /
COPY --from=build /bpftrace/build/src/bpftrace /usr/bin/bpftrace
```
or similar.

File: pkg/bpftrace/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+RUN git clone ${GITREPO} /bpftrace \
+    && cd /bpftrace/ \ 
+    && mkdir -p /bpftrace/build \
+    && git checkout ${COMMIT} \
+    && chmod +x /bpftrace/docker/build.sh \
+    && /bpftrace/docker/build.sh /bpftrace/build Release
+
+FROM scratch
+WORKDIR /
+COPY --from=build /bpftrace/build/src/bpftrace /bpftrace
```
Comment (rn) [inline pkg/bpftrace/Dockerfile]: Please copy this to a standard location, like `/usr/bin`

File: pkg/bpftrace/build.yml [hunk 13]
```diff
@@ -0,0 +1,8 @@
+image: bpftrace
+network: true
+arches: 
+  - amd64
+config:
```
Comment (rn) [inline pkg/bpftrace/build.yml]: This can be dropped if if the main consumption model is to drop the binary into the root filesystem

File: test/cases/040_packages/002_bpftrace/test.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (rn) [reply test/cases/040_packages/002_bpftrace/test.yml]: This should be `linuxkit/bpftrace:<hash>`. To avoid the chicken and egg, I will compile and push the package once and restart the CI.

File: pkg/bpftrace/Dockerfile (hunk 10 above)
Comment (simar7) [reply pkg/bpftrace/Dockerfile]: Hm – not sure about that. I seem to think this will be fine as I was able to run the compiled packaged but is there any other way I could figure that out?

File: pkg/bpftrace/Dockerfile (hunk 11 above)
Comment (simar7) [reply pkg/bpftrace/Dockerfile]: ACK. Makes sense.

File: pkg/bpftrace/Dockerfile (hunk 12 above)
Comment (simar7) [reply pkg/bpftrace/Dockerfile]: ACK.

**Comment by simar7** [comment]
Thanks for looking into this @rn. I've fixed the PR with the suggestions made.
